addons/src/ao_colortip.c

```diff
@@ -0,0 +1,397 @@
+
+struct _AoColorTip
+{
+	GObject parent;
+};
+
+struct _AoColorTipClass
+{
+	GObjectClass parent_class;
+};
+
+struct _AoColorTipPrivate
+{
+	gboolean enable_colortip;
+	gboolean enable_double_click_color_chooser;
+};
+
+enum
+{
+	PROP_0,
+	PROP_ENABLE_COLORTIP,
+	PROP_ENABLE_DOUBLE_CLICK_COLOR_CHOOSER,
+};
+
+G_DEFINE_TYPE(AoColorTip, ao_color_tip, G_TYPE_OBJECT)
+
+#  define SSM(s, m, w, l) scintilla_send_message (s, m, w, l)
+
+
+/* Find a color value (short or long form, e.g. #fff or #ffffff) in the
+   given string. position must be inside the found color or at maximum
+   maxdist bytes in front of it (in front of the start of the color value)
+   or behind it (behind the end of the color value). */
+static gint contains_color_value(gchar *string, gint position, gint maxdist)
+{
+	gchar *start;
+	gint end, value, offset, color = -1;
+	guint length;
+
+	start = strchr(string, '#');
+	if (start == NULL)
+	{
+		return color;
+	}
+	end = start - string + 1;
+	while (g_ascii_isxdigit (string[end]))
+	{
+		end++;
+	}
+	end--;
+	length = &(string[end]) - start + 1;
+
+	if (maxdist != -1)
+	{
+		offset = start - string + 1;
+		if (position < offset &&
+		    (offset - position) > maxdist)
+		{
+			return color;
+		}
+
+		offset = end;
+		if (position > offset &&
+		    (position - offset) > maxdist)
+		{
+			return color;
+		}
+	}
+
+
+	if (length == 4)
+	{
+		start++;
+		color = (g_ascii_xdigit_value (*start) << 4)
+				| g_ascii_xdigit_value (*start);
+
+		start++;
+		value = (g_ascii_xdigit_value (*start) << 4)
+				| g_ascii_xdigit_value (*start);
+		color += value << 8;
+
+		start++;
+		value = (g_ascii_xdigit_value (*start) << 4)
+				| g_ascii_xdigit_value (*start);
+		color += value << 16;
+	}
+	else if (length == 7)
+	{
+		start++;
+		color = (g_ascii_xdigit_value (start[0]) << 4)
+				| g_ascii_xdigit_value (start[1]);
+
+		start += 2;
+		value = (g_ascii_xdigit_value (start[0]) << 4)
+				| g_ascii_xdigit_value (start[1]);
+		color += value << 8;
+
+		start += 2;
+		value = (g_ascii_xdigit_value (start[0]) << 4)
+				| g_ascii_xdigit_value (start[1]);
+		color += value << 16;
+	}
+
+	return color;
+}
+
+
+static gint get_color_value_at_current_doc_position(void)
+{
+	gchar *subtext;
+	gint start, end, pos, max, color = -1;
+	GeanyDocument *doc = document_get_current();
+
+	g_return_val_if_fail(doc != NULL, FALSE);
+
+	/* Is position valid? */
+	pos = sci_get_current_position(doc->editor->sci);
+	if (pos < 0)
+	{
+		return color;
+	}
+	max = SSM(doc->editor->sci, SCI_GETTEXTLENGTH, 0, 0);
+
+	/* Calculate range */
+	start = pos;
+	if (start >= 7)
+	{
+		start -= 7;
+	}
+	else
+	{
+		start = 0;
+	}
+	end = pos + 7;
+	if (end > max)
+	{
+		end = max;
+	}
+
+	/* Get text in range and examine it */
+	subtext = sci_get_contents_range(doc->editor->sci, start, end);
+	if (subtext != NULL)
+	{
+		pos = pos - start;
+		color = contains_color_value(subtext, pos, 1);
+		g_free(subtext);
+	}
+
+	return color;
+}
```

Comment:
any reason why you don't simply use `editor_get_word_at_pos()`?  Maybe something similar to this or whatnot.

```c
gchar *word = editor_get_word_at_pos(doc->editor, -1, "0123456789abcdefABCDEF");

if (word)
{
    gint color = -1;

    switch (strlen (color))
    {
        case 3:
            color = ((g_ascii_xdigit_value(word[0]) * 0x11) << 16 |
                     (g_ascii_xdigit_value(word[1]) * 0x11) << 8 |
                     (g_ascii_xdigit_value(word[2]) * 0x11) << 0);
            break;
        case 6:
            color = (g_ascii_xdigit_value(word[0]) << 20 |
                     g_ascii_xdigit_value(word[1]) << 16 |
                     g_ascii_xdigit_value(word[2]) << 12 |
                     g_ascii_xdigit_value(word[3]) << 8 |
                     g_ascii_xdigit_value(word[4]) << 4 |
                     g_ascii_xdigit_value(word[5]) << 0);
            break;
        default:
            /* invalid color or other format */
            break;
    }
    // …
````

Reply:
Thanks, used your code snippet.